### ISSUE TITLE
Warnings with ruby 2.5.0

### ISSUE THREAD

**Issue by Dorian** [ruby/rake#248]
I get lots of warnings when running rake with ruby 2.5.0

```
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_utils.rb:10: warning: already initialized constant FileUtils::RUBY
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils.rb:10: warning: previous definition of RUBY was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_utils.rb:109: warning: already initialized constant FileUtils::LN_SUPPORTED
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils.rb:109: warning: previous definition of LN_SUPPORTED was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:17: warning: already initialized constant Rake::FileUtilsExt::DEFAULT
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:17: warning: previous definition of DEFAULT was here
WARNING: Possible conflict with Rake extension: String#ext already exists
WARNING: Possible conflict with Rake extension: String#pathmap already exists
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:44: warning: already initialized constant Rake::FileList::ARRAY_METHODS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:44: warning: previous definition of ARRAY_METHODS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:47: warning: already initialized constant Rake::FileList::MUST_DEFINE
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:47: warning: previous definition of MUST_DEFINE was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:51: warning: already initialized constant Rake::FileList::MUST_NOT_DEFINE
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:51: warning: previous definition of MUST_NOT_DEFINE was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:55: warning: already initialized constant Rake::FileList::SPECIAL_RETURN
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:55: warning: previous definition of SPECIAL_RETURN was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:61: warning: already initialized constant Rake::FileList::DELEGATING_METHODS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:61: warning: previous definition of DELEGATING_METHODS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:86: warning: already initialized constant Rake::FileList::GLOB_PATTERN
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:86: warning: previous definition of GLOB_PATTERN was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:381: warning: already initialized constant Rake::FileList::DEFAULT_IGNORE_PATTERNS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:381: warning: previous definition of DEFAULT_IGNORE_PATTERNS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/file_list.rb:387: warning: already initialized constant Rake::FileList::DEFAULT_IGNORE_PROCS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_list.rb:387: warning: previous definition of DEFAULT_IGNORE_PROCS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:3: warning: already initialized constant Rake::VERSION
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:3: warning: previous definition of VERSION was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::MAJOR
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of MAJOR was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::MINOR
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of MINOR was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::BUILD
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of BUILD was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:6: warning: already initialized constant Rake::Version::OTHER
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:6: warning: previous definition of OTHER was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/version.rb:8: warning: already initialized constant Rake::Version::NUMBERS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/version.rb:8: warning: previous definition of NUMBERS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/linked_list.rb:110: warning: already initialized constant Rake::LinkedList::EMPTY
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/linked_list.rb:110: warning: previous definition of EMPTY was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/scope.rb:41: warning: already initialized constant Rake::Scope::EMPTY
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/scope.rb:41: warning: previous definition of EMPTY was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/promise.rb:12: warning: already initialized constant Rake::Promise::NOT_SET
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/promise.rb:12: warning: previous definition of NOT_SET was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/application.rb:13: warning: already initialized constant Rake::CommandLineOptionError
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:13: warning: previous definition of CommandLineOptionError was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/application.rb:41: warning: already initialized constant Rake::Application::DEFAULT_RAKEFILES
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:41: warning: previous definition of DEFAULT_RAKEFILES was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/task_arguments.rb:108: warning: already initialized constant Rake::EMPTY_TASK_ARGS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task_arguments.rb:108: warning: previous definition of EMPTY_TASK_ARGS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/invocation_chain.rb:55: warning: already initialized constant Rake::InvocationChain::EMPTY
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/invocation_chain.rb:55: warning: previous definition of EMPTY was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/early_time.rb:21: warning: already initialized constant Rake::EARLY
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/early_time.rb:21: warning: previous definition of EARLY was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/late_time.rb:17: warning: already initialized constant Rake::LATE
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/late_time.rb:17: warning: previous definition of LATE was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:4: warning: already initialized constant Rake::Backtrace::SYS_KEYS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:4: warning: previous definition of SYS_KEYS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:5: warning: already initialized constant Rake::Backtrace::SYS_PATHS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:5: warning: previous definition of SYS_PATHS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:8: warning: already initialized constant Rake::Backtrace::SUPPRESSED_PATHS
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:8: warning: previous definition of SUPPRESSED_PATHS was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:12: warning: already initialized constant Rake::Backtrace::SUPPRESSED_PATHS_RE
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:12: warning: previous definition of SUPPRESSED_PATHS_RE was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake/backtrace.rb:16: warning: already initialized constant Rake::Backtrace::SUPPRESS_PATTERN
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/backtrace.rb:16: warning: previous definition of SUPPRESS_PATTERN was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake.rb:70: warning: already initialized constant FileList
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake.rb:70: warning: previous definition of FileList was here
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0/lib/rake.rb:71: warning: already initialized constant RakeFileUtils
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake.rb:71: warning: previous definition of RakeFileUtils was here
```

**Comment by gersonazgo** [comment]
Same problem here.

**Comment by hsbt** [comment]
It seems RVM issue. 

```
/Users/d/.rvm/gems/ruby-2.5.0@global/gems/rake-12.3.0
/Users/d/.rvm/rubies/ruby-2.5.0/lib/ruby/gems/2.5.0/gems/rake-12.3.0
```

You should remove duplicated path from above list.

**Comment by mpapis** [comment]
duplicate paths should never be a problem as long as `$LOAD_PATH` is used, a simple test:
`a/test.rb` & `b/test.rb`:
```ruby
class Test
  TEST = "test"
end
```
and run:
```bash
# ruby -Ia -r test -Ib -rtest -e 'puts Test::TEST'
test
# ruby -r./a/test.rb -r./b/test.rb -e 'puts Test::TEST'                                                                                                                                                        
/home/mpapis/tmp/test_load_path_rb/b/test.rb:2: warning: already initialized constant Test::TEST                                                                                                                                                                              
/home/mpapis/tmp/test_load_path_rb/a/test.rb:2: warning: previous definition of TEST was here                                                                                                                                                                                 
test                                                                                                                                                                                                                                                                          
```
Somewhere there is a relative require, it should be avoided, especially in gems (well I know of one use case, but that's in a rubygems plugin that's triggered outside of normal flow).

**Comment by mpapis** [comment]
also not rake issue, the problem is triggered with `bootsnap`, using `rails new myapp --skip-bootsnap` it works all fine

**Comment by bbugh** [comment]
This seems to be the source of the issue: https://github.com/Shopify/bootsnap/issues/134

**Comment by hsbt** [comment]
@mpapis Thanks for your clarification. I apologized for my wrong comment( https://github.com/ruby/rake/issues/248#issuecomment-360766271 )

**Comment by wellington1993** [comment]
Hi, I removed `require 'bootsnap/setup'` from `boot.rb` and worked for me.

**Comment by mpapis** [comment]
@hsbt no problem, I got used to RVM being blamed for everything ;)

**Comment by concept47** [comment]
Hi guys. I'm a bit unclear about the resolution here? Its not a rake issue, and its not an rvm issue? So is it a bootsnap issue?

**Comment by maca** [comment]
Seems to be a rack/bundler issue. 
Only when using ruby 2.5 I get this warnings.

I am using `Bundler.require(:default, ENV["RACK_ENV"] ||= "development")` to load all gems in an environment file.

When running `$ bundle exec rake -T` I get no warnings
I can get rid of the warnings by skipping the requirement of rake in the Gemfile: `gem 'rake', require: false`

**Comment by concept47** [comment]
Thank you @maca thats very interesting to note.

**Comment by Lawrence-Sproul** [comment]
Just update bootsnap. I resolved this by putting version 1.3.1 in my Gemfile.
`gem 'bootsnap', '>= 1.3.1', require: false`
